File: packages/c#/RemoteProcessClient.cs
using System.Net.Sockets;
using System.IO;
using System.Collections.Generic;
using Newtonsoft.Json;

namespace Com.CodeGame.CodeBall2018.DevKit.CSharpCgdk
{
    public sealed class RemoteProcessClient
    {
        TcpClient client;
        StreamReader reader;
        StreamWriter writer;
        public RemoteProcessClient(string host, int port)
        {
            client = new TcpClient(host, port) { NoDelay = true };
            reader = new StreamReader(client.GetStream());
            writer = new StreamWriter(client.GetStream());

            writer.WriteLine("json");
            writer.Flush();
        }

        public Model.Game ReadGame()
        {
            string line = reader.ReadLine();
            if (line == null || line.Length == 0)
            {
                return null;
            }
            else
            {
                return JsonConvert.DeserializeObject<Model.Game>(line);
            }
        }

        public Model.Rules ReadRules()
        {
            string line = reader.ReadLine();
            if (line == null || line.Length == 0)
            {
                return null;
            }
            else
            {
                return JsonConvert.DeserializeObject<Model.Rules>(line);
            }
        }

        public void Write(IDictionary<int, Model.Action> actions, string custom_rendering)
        {
            writer.WriteLine(JsonConvert.SerializeObject(actions) + "|" + custom_rendering + "\n<end>");
            writer.Flush();
        }

        public void WriteToken(string token)
        {
            writer.WriteLine(token);
            writer.Flush();
        }
    }
}

File: packages/python3/RemoteProcessClient.py
import socket
import io
import json
from model import *


class JSONEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Action):
            return obj.__dict__
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)


class RemoteProcessClient:
    def __init__(self, host, port):
        self.socket = socket.socket()
        self.socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, True)
        self.socket.connect((host, port))
        self.reader = io.BufferedReader(
            socket.SocketIO(self.socket, 'r'))
        self.writer = self.socket.makefile('w')
        self.writer.write("json\n")
        self.writer.flush()

    def read_game(self):
        line = self.reader.readline()
        return Game(json.loads(line)) if line else None

    def read_rules(self):
        line = self.reader.readline()
        return Rules(json.loads(line)) if line else None

    def write(self, actions, custom_rendering):
        self.writer.write(json.dumps(
            actions, separators=(',', ':'), cls=JSONEncoder))
        self.writer.write('|')
        self.writer.write(custom_rendering)
        self.writer.write('\n<end>\n')
        self.writer.flush()

    def write_token(self, token):
        self.writer.write(token)
        self.writer.write('\n')
        self.writer.flush()
